include/counter.h

```diff
@@ -22,35 +22,133 @@
 #include <zephyr/types.h>
 #include <stddef.h>
 #include <device.h>
+#include <stdbool.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-typedef void (*counter_callback_t)(struct device *dev, void *user_data);
+struct counter_alarm_callback;
+
+/** @brief Alarm callback
+ *
+ * @param dev     Pointer to the device structure for the driver instance.
+ * @param cb      Original structure owning this handler.
+ * @param counter Counter value that triggered the callback.
+ */
+typedef void (*counter_alarm_callback_handler_t)(struct device *dev,
+					 struct counter_alarm_callback *cb,
```

Comment:
I find this naming a bit confusing. To me, a callback is just a function to be called when certain circumstance occurs, not a whole structure describing when/how this function should be called. And a callback handler sounds rather like an additional piece of code to handle the callback itself in some fancy way, get additional data needed to execute it etc. Couldn't we go with `counter_alarm` structure and associated `counter_alarm_callback`?

Reply:
Or better `counter_alarm_cfg`, as suggested by @MaureenHelm, to keep consistency with watchdog.